compiler/noirc_evaluator/src/ssa.rs

```diff
@@ -149,6 +115,69 @@ pub(crate) fn optimize_into_acir(
     Ok(ArtifactsAndWarnings(artifacts, ssa_level_warnings))
 }
 
+fn optimize_ssa(builder: SsaBuilder, inliner_aggressiveness: i64) -> Result<Ssa, RuntimeError> {
+    let builder = builder
+        .run_pass(Ssa::defunctionalize, "After Defunctionalization:")
+        .run_pass(Ssa::remove_paired_rc, "After Removing Paired rc_inc & rc_decs:")
+        .run_pass(Ssa::separate_runtime, "After Runtime Separation:")
+        .run_pass(Ssa::resolve_is_unconstrained, "After Resolving IsUnconstrained:")
+        .run_pass(|ssa| ssa.inline_functions(inliner_aggressiveness), "After Inlining:")
+        .run_pass(
+            |ssa| ssa.inline_const_brillig_calls(inliner_aggressiveness),
+            "After Inlining Const Brillig Calls:",
+        );
+    let ssa = optimize_ssa_after_inline_const_brillig_calls(
+        builder,
+        inliner_aggressiveness,
+        true, // inline functions with no predicates
+    )?;
+    Ok(ssa)
```

Comment:
Is there a reason to break up `optimize_ssa` this way? Why is "after inline cost brillig calls" a significant point?

Reply:
Ah found it. Maybe some docstrings could explain, like help decide if a new pass should be before or after this?

Reply:
Oh, it's mainly because at this point the other optimizations already happened, so to avoid redoing them again (and I think the inlining pass might break things too).

Should I add a comment for this somewhere?

Reply:
I would mention it on `optimize_ssa_after_inline_const_brillig_calls` that it's used to tentatively try to optimise away functions by the `.inline_const_brillig_calls` itself. 

I keep thinking that it would be nice if the `inline_const_brillig_calls` didn't have to know exactly which steps are following it. The `SsaBuilder` isn't an abstract one that one can pass around and customise, it's tied to an `Ssa` instance and the caller has to drive the passes, but ostensibly `inline_cost_brillig_calls` could get a closure that takes an `Ssa` and runs what we want on it, and then this flow would be visible here.

Reply:
Something like:
```rust
fn optimize_ssa(builder: SsaBuilder, inliner_aggressiveness: i64) -> Result<Ssa, RuntimeError> {
    fn stage1(builder: SsaBuilder, inliner_aggressiveness: i64) -> Result<SsaBuilder, RuntimeError> {
        Ok(builder
        .run_pass(Ssa::defunctionalize, "After Defunctionalization:")
        .run_pass(Ssa::remove_paired_rc, "After Removing Paired rc_inc & rc_decs:")
        .run_pass(Ssa::separate_runtime, "After Runtime Separation:")
        .run_pass(Ssa::resolve_is_unconstrained, "After Resolving IsUnconstrained:")
        .run_pass(|ssa| ssa.inline_functions(inliner_aggressiveness), "After Inlining:")
        .run_pass(
            |ssa| ssa.inline_const_brillig_calls(|ssa| {
                let builder = SsaBuilder { ssa, print_stuff: false };
                stage2(builder, false).map(|b| b.finish())
            }),
            "After Inlining Const Brillig Calls:",
        ))
    }

    fn stage2(builder: SsaBuilder, inliner_aggressiveness: i64, inline_functions_with_no_predicates) -> Result<SsaBuilder, RuntimeError> {
        Ok(builder
        // Run mem2reg with the CFG separated into blocks
        .run_pass(Ssa::mem2reg, "After Mem2Reg (1st):")
        .run_pass(Ssa::simplify_cfg, "After Simplifying (1st):")
        .run_pass(Ssa::as_slice_optimization, "After `as_slice` optimization")
        .try_run_pass(
            Ssa::evaluate_static_assert_and_assert_constant,
            "After `static_assert` and `assert_constant`:",
        )?
        .try_run_pass(Ssa::unroll_loops_iteratively, "After Unrolling:")?
        .run_pass(Ssa::simplify_cfg, "After Simplifying (2nd):")
        .run_pass(Ssa::flatten_cfg, "After Flattening:")
        .run_pass(Ssa::remove_bit_shifts, "After Removing Bit Shifts:")
        // Run mem2reg once more with the flattened CFG to catch any remaining loads/stores
        .run_pass(Ssa::mem2reg, "After Mem2Reg (2nd):")
        // Run the inlining pass again to handle functions with `InlineType::NoPredicates`.
        // Before flattening is run, we treat functions marked with the `InlineType::NoPredicates` as an entry point.
        // This pass must come immediately following `mem2reg` as the succeeding passes
        // may create an SSA which inlining fails to handle.
        .run_pass(
            |ssa| if inline_functions_with_no_predicates {
                ssa.inline_functions_with_no_predicates(inliner_aggressiveness)
            } else {
                ssa
            },
            "After Inlining:",
        )
        .run_pass(Ssa::remove_if_else, "After Remove IfElse:")
        .run_pass(Ssa::fold_constants, "After Constant Folding:")
        .run_pass(Ssa::remove_enable_side_effects, "After EnableSideEffectsIf removal:")
        .run_pass(Ssa::fold_constants_using_constraints, "After Constraint Folding:")
        .run_pass(Ssa::dead_instruction_elimination, "After Dead Instruction Elimination:")
        .run_pass(Ssa::simplify_cfg, "After Simplifying:")
        .run_pass(Ssa::array_set_optimization, "After Array Set Optimizations:"))
    }

    Ok(stage1(builder, true)?.finish())
}
```

Reply:
Maybe add a `run_pass_if` variant to avoid printing if a step is turned off (not that it would print in this particular case, but just to not have to break up the chain).

Reply:
> didn't have to know exactly which steps are following it

To illustrate this point, currently it looks something like this:

```rust
fn optimise(builder: SsaBuilder) -> Ssa {
   foo(builder)
}
fn foo(builder: SsaBuiler) -> Ssa {
    builder.run(Ssa::a);
    builder.run(Ssa::b);
    builder.run(Ssa::c);
    bar(builder)
}
fn bar(builder: SsaBuiler) -> Ssa {
    builder.run(Ssa::d);
    builder.run(Ssa::e);
    baz(builder)
}
fn baz(builder: SsaBuiler) -> Ssa {
    builder.run(Ssa::f);
    builder.finish()
}

// Individual modules weave together all methods into Ssa:
impl Ssa {
  fn a(self) -> Self { todo!(); }
  fn b(self) -> Self { todo!(); }
  fn c(self) -> Self { todo!(); let ssa = bar(self); todo!() }
  fn d(self) -> Self { todo!(); let ssa = baz(self); todo!() }
  fn e(self) -> Self { todo!(); }
  fn f(self) -> Self { todo!(); }
}
```

That is, `c` and `d` know to call `bar` and `baz` to do "the rest" of the optimisations, but  if someone was to refactor the top level functions they would have to go an update `c` and `d` to call the correct entry points, make sure there is one, and they would be going up and down in the code to follow the chain of calls. For example consider the following refactorings: 
* moving `c` into `bar` would cause an infinite loop
* moving `d` into `foo` would change what comes after `c` and might invalidate its assumptions about what happens to the Ssa

By passing in a closure to `c` and `d`, the sequencing of the passes would all be visible in the top level builder. The composer would need to understand what `c` needs, but it could be documented and it would be obviously there, rather than a method named `passes_after_c` and e.g. `passes_after_d_but_before_f` which sound a bit more accidental.

Reply:
I guess another option would be to put all passes into an ordered map and make it possible to resume from anywhere by ID.

Reply:
I'll try to refactor this to one of the options you suggested.

Reply:
I just tried it but it's complex because of the `#[tracing::instrument(level = "trace", skip(self))]` annotation on top of asses: we can't pass closures or `fn` pointers there as they are not `Debug`.

I'm also thinking that the current way is okay: we are running this `inline_const_brillig_calls` optimization so it makes sense to only run passes that comes after it, hence the name I chose.

> I keep thinking that it would be nice if the inline_const_brillig_calls didn't have to know exactly which steps are following it

It doesn't: it just runs the passes that comes after it, but it still doesn't know which are those.

Alternatively... I could just call individual passes when optimizing this function. It's actually what I originally had, but I didn't want us to have to remember to add new optimizations in two places. I guess that's the tradeoff.

Reply:
Thanks for trying. Yeah, I admit it's rather philosophical. I mean now that there is a single deviated "after_inlining_const_brillig" it's easy to say that it doesn't actually know, but in another way it knows that this function is there for it to call; the only way I can say is that if there were more of them, and they weren't so deftly named after the caller itself, it would be more obvious that it has to know which one to pick, rather than being given by the caller. Something like Dependency Injection. Never mind, there is only one :)